homeassistant/components/home_connect/coordinator.py

```diff
@@ -0,0 +1,227 @@
+                                else:
+                                    settings[setting_key] = GetSetting(
+                                        setting_key, event.value
+                                    )
+                            await self._call_event_listener(event_message)
+
+                        case EventType.EVENT:
+                            await self._call_event_listener(event_message)
+
+            except (EventStreamInterruptedError, HomeConnectRequestError) as error:
+                _LOGGER.debug(
+                    "Non-breaking error (%s) while listening for events,"
+                    " continuing in 30 seconds",
+                    type(error).__name__,
+                )
+                await asyncio.sleep(EVENT_STREAM_RECONNECT_DELAY)
+            except HomeConnectApiError as error:
+                _LOGGER.error("Error while listening for events: %s", error)
+                break
+
+    async def _call_event_listener(self, event_message: EventMessage):
+        """Call listener for event."""
+        listeners = self._home_appliances_event_listeners.get(event_message.ha_id, {})
+        if not listeners:
+            return
+        for event in event_message.data.items:
+            if event.key is EventKey.UNKNOWN:
+                continue
+            for listener in listeners.get(event.key, []):
+                await listener(event)
+
+    def add_home_appliances_event_listener(
+        self,
+        ha_id: str,
+        event_key: EventKey,
+        cb: Callable[[Event], Coroutine[Any, Any, None]],
+    ) -> None:
+        """Register a event listener for a specific home appliance and event key."""
+        self._home_appliances_event_listeners.setdefault(ha_id, {}).setdefault(
+            event_key, []
+        ).append(cb)
+
+    def delete_home_appliances_event_listener(
+        self,
+        ha_id: str,
+        event_key: EventKey,
+        cb: Callable[[Event], Coroutine[Any, Any, None]],
+    ) -> None:
+        """Deregister event listener for a specific home appliance and event key."""
+        if (
+            ha_id in self._home_appliances_event_listeners
+            and event_key in self._home_appliances_event_listeners[ha_id]
+        ):
+            appliance_listeners = self._home_appliances_event_listeners[ha_id]
+            if cb in (
+                listeners := self._home_appliances_event_listeners[ha_id][event_key]
+            ):
+                listeners.remove(cb)
+            if not listeners:
+                self._home_appliances_event_listeners[ha_id].pop(event_key)
+            if not appliance_listeners:
+                self._home_appliances_event_listeners.pop(ha_id)
+
+    async def _async_update_data(self) -> dict[str, HomeConnectApplianceData]:
+        """Fetch data from Home Connect."""
+        try:
+            appliances = await self.client.get_home_appliances()
+        except HomeConnectError as error:
+            raise UpdateFailed(
```

Comment:
If we get a unauthorized error from the API here we want to start a reauth flow by raising `ConfigEntryAuthFailed`. This can be added in a follow up PR.